stan/math/fwd/functor/operands_and_partials.hpp

```diff
@@ -2,33 +2,201 @@
 #define STAN_MATH_FWD_META_OPERANDS_AND_PARTIALS_HPP
 
 #include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/meta.hpp>
 #include <stan/math/prim/fun/size.hpp>
+#include <stan/math/fwd/core/fvar.hpp>
+#include <stan/math/fwd/meta.hpp>
 #include <stan/math/prim/functor/broadcast_array.hpp>
 #include <stan/math/prim/functor/operands_and_partials.hpp>
-#include <stan/math/fwd/core/fvar.hpp>
+#include <stan/math/prim/functor/apply.hpp>
+#include <stan/math/prim/functor/for_each.hpp>
 #include <vector>
 
 namespace stan {
 namespace math {
 namespace internal {
 
-template <typename Dx>
-class ops_partials_edge<Dx, fvar<Dx>> {
+/**
+ * End of recursion for summing `.dx()` for `fvar<T>` ops and partials.
+ */
+static constexpr auto sum_dx() { return static_cast<double>(0.0); }
+
+/**
+ * End of recursion for summing `.dx()` for `fvar<T>` ops and partials.
+ * @tparam T1 a type with a `.dx()` method.
+ * @param a an edge from `operands_and_partials_impl`
+ */
+template <typename T1>
+inline auto sum_dx(T1& a) {
+  return a.dx();
+}
+
+/**
+ * Accumulate the `.dx()` from each of the edges in `ops_and_partials_impl`.
+ * @tparam T1 a type with a `.dx()` method.
+ * @tparam T2 a type with a `.dx()` method.
+ * @tparam Types a type with a `.dx()` method.
+ * @param a an edge from `operands_and_partials_impl`
+ * @param b an edge from `operands_and_partials_impl`
+ * @param args edges from `operands_and_partials_impl`
+ */
+template <typename T1, typename T2, typename... Types>
+inline auto sum_dx(T1& a, T2& b, Types&... args) {
+  return a.dx() + b.dx() + sum_dx(args...);
+}
```

Comment:
Little cleaner to replace this recursion with a fold expression:
```cpp

template<typename... Types>
auto sum_dx(const Types&... args) {
  return_type_t<Types...> result = {};
  (void)std::initializer_list<int>{ (result += v.dx(), 0)... };
  return result;
}
```

Reply:
This is a little confusing imo I'd rather using the recursive parameter expansion to keep it a bit simpler. Inlining should roll out this call for the 2-5 sized argument sizes we normally have

Reply:
Fair enough!